pkg/encrypt/dare.go

```diff
@@ -0,0 +1,87 @@
+/*
+ * Minio Go Library for Amazon S3 Compatible Cloud Storage (C) 2017 Minio, Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package encrypt
+
+import (
+	"crypto/hmac"
+	"crypto/rand"
+	"crypto/sha256"
+	"encoding/base64"
+	"errors"
+	"io"
+	"io/ioutil"
+
+	"github.com/minio/sio"
+)
+
```

Comment:
Can you add a comment in this file saying that we are satisfying Cipher interface ?

Reply:
I'm not sure if I understand your comment correctly. Are you requesting docs for the `dareHmacSha256` type?
I think something like
```
// dareHmacSha256 implements encrypt.Cipher with the DARE format
// and HMAC-SHA256 as encryption key derivation function.
type dareHmacSha256 [32]byte
```

Reply:
is reasonable?!

Reply:
Actually I was talking about the whole file and what it is  for.. because if we don't have a priori knowledge about the design, we can get lost when we read this file, it doesn't clearly say this file satisfy Cipher interface.

Reply:
Hm - I'm not sure about this... This style (special implementations in separate files) is kind of well known in the Go world - I would expect separated implementations in more or less this way (standard library)...
On the other side a comment is no big deal. I'm just not sure whether we should start something like this...

Reply:
> This style (special implementations in separate files) is kind of well known in the Go world

yes.. but if you don't know that this file implements Cipher, how much time it takes for you to know that ? e.g. is there an emacs shortcut that can help to do that ?

Reply:
That's the advantage of Go interfaces - that types can implement interfaces without knowing that the interface exists. The downside is that this relationship is not explicit anymore.
If you want to make it explicit (and enforced by the compiler) we can add:
`var _ Cipher = (*dareHmacSha256)(nil)`
That would be more explicit that a comment.

Reply:
Sorry @aead, it looks like you already did this (not sure when)

```
// dareHmacSha256 implements encrypt.Cipher with the DARE format
// and HMAC-SHA256 as encryption key derivation function.
```

For me, this is sufficient.

Generally, in a package with few files, it is not problem, but when there are many files in the package, it will be hard to know that a type satisfies an interface when you read the code the first time. That's my point.

So the current state already looks good.

Reply:
`HEAD~1` 😉 
Completely agree - can be hard in large packages.